### ISSUE TITLE
Device lost on manager reload

### ISSUE THREAD

**Issue by matijaGP** [systemd/systemd#19502]
Hello,

with our update to v247/v248 we were bitten by udev bind rework. I believe I have case which is a design issue:

We have platform device driver which deals with several instances of fan devices on our boards. We use change event in driver to setup systemd devices and aliases, here is how our rule looks like:

```
ACTION=="change", KERNEL=="*fsm-fan*", SUBSYSTEM=="platform", ENV{KERNEL}="%k", \
    TAG+="systemd", ENV{SYSTEMD_ALIAS}="/dev/ddal/fan/%E{INDEX}", \
    RUN+="/usr/lib/udev/fan.sh", RUN+="/usr/lib/udev/hwt-fan-fsm.sh"
```

INDEX is set to uevent in driver, and is used to select instance of operation.

I went through series which did rework in the area (https://github.com/systemd/systemd/pull/16853/commits). In our case, we are broken by the fact that CURRENT_TAG gets cleared on bind event. Further, bind event carries not enough information to setup alias, therefore our alias gets broken as well.

Here is how it looks like on loading the module:

```
UDEV  [4835.918251] change   /devices/platform/soc/soc:picasso/3080005010.fsm-fan (platform)
ACTION=change
DEVPATH=/devices/platform/soc/soc:picasso/3080005010.fsm-fan
SUBSYSTEM=platform
CHANGE_ACTION=init
INDEX=1
DRIVER=fsm-fan
OF_NAME=fsm-fan
OF_FULLNAME=/soc/picasso/fsm-fan@00005010,1
OF_COMPATIBLE_0=nsn,fan-fpga-picasso
OF_COMPATIBLE_N=1
MODALIAS=of:Nfsm-fanT(null)Cnsn,fan-fpga-picasso
SEQNUM=2034
USEC_INITIALIZED=12045366
ID_PATH=platform-3080005010.fsm-fan
ID_PATH_TAG=platform-3080005010_fsm-fan
KERNEL=3080005010.fsm-fan
SYSTEMD_ALIAS=/dev/ddal/fan/1
TAGS=:systemd:
CURRENT_TAGS=:systemd:
```

```
UDEV  [4835.920457] bind     /devices/platform/soc/soc:picasso/3080005010.fsm-fan (platform)
ACTION=bind
DEVPATH=/devices/platform/soc/soc:picasso/3080005010.fsm-fan
SUBSYSTEM=platform
DRIVER=fsm-fan
OF_NAME=fsm-fan
OF_FULLNAME=/soc/picasso/fsm-fan@00005010,1
OF_COMPATIBLE_0=nsn,fan-fpga-picasso
OF_COMPATIBLE_N=1
MODALIAS=of:Nfsm-fanT(null)Cnsn,fan-fpga-picasso
SEQNUM=2035
USEC_INITIALIZED=12045366
ID_PATH=platform-3080005010.fsm-fan
ID_PATH_TAG=platform-3080005010_fsm-fan
TAGS=:systemd:
```

As you can see, first we get change event with everything set up. Bind follows up, but it has cleared several properties, CURRENT_TAGS and SYSTEMD_ALIAS.

Without reloading, everything would work, however on reload, 'main' device gets into state Dead, due to this one https://github.com/poettering/systemd/commit/242c1c075aa284c8a8657c5aca36147f528146ba. During the serialize, condition is not met due to fact that latest bind event removed systemd tag from current, so device doesnt get deserialized. Alias device as well.

Even if above comit is removed (for the sake of excercise, I dont see problem with implementation, but rather idea), alias device would still be left dead.

At the moment, I have solved this by providing additional rule:

```
ACTION=="bind", KERNEL=="*fsm-fan*", SUBSYSTEM=="platform", ENV{KERNEL}="%k", \
    TAG+="systemd" IMPORT{db}="CURRENT_TAGS" IMPORT{db}="SYSTEMD_ALIAS"
```

However, I see it more as a workaround.

Could you please tell me, what is your view on this? I believe use case is quite valid one, not to have subsequent bind event removing properties from previous change. Controlling order of those would also be quite tricky as the most logic place to put generation of change is within driver probe function, while bind is fired once probe is finished.

Kernel version is 5.4, systemd v247, v248. We are updating from v239.

Thanks,

Matija

**Comment by poettering** [comment]
i don't follow? you have a "change" uevent with an INDEX property? But you rule doesn't use it except for SYSTEMD_ALIAS?

what are you precisely trying to do?

If these are multiple fans, why aren#t they exposed as multiple devices? Are you shoehorning multiple devices into a single sysfs device and surprised this falls apart?

**Comment by matijaGP** [comment]
INDEX is used in setting up alias, and also for creating usable links to the device. In this particular instance, we have two devices being instantiated, each fan being represented by its own device and sysfs:

```
  dev-ddal-fan-0.device                                               loaded    active   plugged   /dev/ddal/fan/0                                                                              
  dev-ddal-fan-1.device                                               loaded    active   plugged   /dev/ddal/fan/1                                                                              
  sys-devices-platform-soc-soc:picasso-3080005000.fsm\x2dfan.device   loaded    active   plugged   /sys/devices/platform/soc/soc:picasso/3080005000.fsm-fan                                     
  sys-devices-platform-soc-soc:picasso-3080005010.fsm\x2dfan.device   loaded    active   plugged   /sys/devices/platform/soc/soc:picasso/3080005010
```

Devices are used further to tailor dependencies, while links for accessing sysfs.

**Comment by poettering** [comment]
> INDEX is used in setting up alias, and also for creating usable links to the device

hmm, are you saying htere's also a SYMLINKS line?

why is the INDEX stuff missing from the bind event?

can't you read the index stuff as sysfs attr too?

If the driver only makes the INDEX stuff available in change events, but not the other types of events or sysfs, then that looks like something to fix.

**Comment by matijaGP** [comment]
symlinking is done from the script, not directly by udev. There are no more rules besides those I posted.

Bind is not under control from driver. It is called after probe returns success, from driver_bound: https://elixir.bootlin.com/linux/v5.4.116/source/drivers/base/dd.c#L369

Only small portion of properties are sent by it. It offers no interface, besides telling "driver is now bound". As such it cannot be used in this scenario.

If one wants to have interface when "add" doesnt satisfy, "change" is way to go as one has freedom to fill environment per needs - as we fill with INDEX. It is same mechanism as using SYSTEMD_READY=0, and then later sending "change" which has it set to 1 (described in systemd device manual page). I believe that mechanism would also be invalidated in similar scenario.

Index cannot be read from sysfs as we need it at the time of rule being evaluated to have alias set. AFAIK, that's not possible to do from script triggered by the rule, it needs to be within the rule.

There is for sure some other way to achieve what we need, one being example I posted, however idea of this ticket is that what is configured in change event needs to be permanent for the object timeline, or until explicitly changed, and not as a side effect of next event. bind is clearing those properties, resulting with vanished device on reload. I understand bind story is complex and it brought problems, but please try to consider this as valid usage and simply scenario which is victim of current handling.

Slight digression - theoretical problem in the area might as well be situation in which e.g. block devices will get bind event after change. Just plugging usb into my laptop (this is other machine from the one in original post) shows that it is change event which is last, and seemingly hosting all necessary information. Imagine bind following up on this, and then manager reload - device might vanish (ofcourse, it might not as it has related mounts, what I'm trying to point is that there could be more trouble with this if events would be reordered or changed in a way that 'empty' event follows useful one).

```
UDEV  [384792.235580] change   /devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/host3/target3:0:0/3:0:0:0/block/sdb/sdb1 (block)
ACTION=change
DEVPATH=/devices/pci0000:00/0000:00:14.0/usb1/1-1/1-1:1.0/host3/target3:0:0/3:0:0:0/block/sdb/sdb1
SUBSYSTEM=block
SYNTH_UUID=1d675a70-62fc-41b8-ab37-77e587bbfac5
SYNTH_ARG_UDISKSSERIAL=3
DEVNAME=/dev/sdb1
DEVTYPE=partition
PARTN=1
SEQNUM=4061
USEC_INITIALIZED=384792096714
ID_VENDOR_ENC=\x20\x20\x20\x20\x20\x20\x20\x20
ID_VENDOR_ID=0930
ID_MODEL=USB_Flash_Memory
ID_MODEL_ENC=USB\x20Flash\x20Memory
ID_MODEL_ID=6544
ID_REVISION=1.00
ID_SERIAL=_USB_Flash_Memory_54B80A3FA797C090E7BA50E8-0:0
ID_SERIAL_SHORT=54B80A3FA797C090E7BA50E8
ID_TYPE=disk
ID_INSTANCE=0:0
ID_BUS=usb
ID_USB_INTERFACES=:080650:
ID_USB_INTERFACE_NUM=00
ID_USB_DRIVER=usb-storage
ID_PATH=pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:0
ID_PATH_TAG=pci-0000_00_14_0-usb-0_1_1_0-scsi-0_0_0_0
ID_PART_TABLE_UUID=72296cec
ID_PART_TABLE_TYPE=dos
ID_FS_UUID=2019-11-16-11-35-08-00
ID_FS_UUID_ENC=2019-11-16-11-35-08-00
ID_FS_BOOT_SYSTEM_ID=EL\x20TORITO\x20SPECIFICATION
ID_FS_VERSION=Joliet Extension
ID_FS_LABEL=Debian_10.2.0_i386_n
ID_FS_LABEL_ENC=Debian\x2010.2.0\x20i386\x20n
ID_FS_TYPE=iso9660
ID_FS_USAGE=filesystem
ID_PART_ENTRY_SCHEME=dos
ID_PART_ENTRY_UUID=72296cec-01
ID_PART_ENTRY_TYPE=0x0
ID_PART_ENTRY_FLAGS=0x80
ID_PART_ENTRY_NUMBER=1
ID_PART_ENTRY_OFFSET=0
ID_PART_ENTRY_SIZE=872448
ID_PART_ENTRY_DISK=8:16
.ID_FS_TYPE_NEW=iso9660
.MM_USBIFNUM=00
UDISKS_IGNORE=0
MAJOR=8
MINOR=17
DEVLINKS=/dev/disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:0-part1 /dev/disk/by-id/usb-_USB_Flash_Memory_54B80A3FA797C090E7BA50E8-0:0-part1 /dev/disk/by-label/Debian\x2010.2.0\x20i386\x20n /dev/disk/by-uuid/2019-11-16-11-35-08-00 /dev/disk/by-partuuid/72296cec-01
TAGS=:systemd:
CURRENT_TAGS=:systemd:
```

**Comment by poettering** [comment]
no sure i follow. typically block devices only become available after the "bind".

is yours an upstream driver? quite frankly, the driver appears quite broken to me if it only passes some information you need to userspace on "change". this is not how this is supposed to work, if you want information that shall have relevance beyond the immediate event.

**Comment by matijaGP** [comment]
Information passing in this context is udev related information. If driver wants to share such info, which would be used in udev rules, I dont see other way. Bind is not under driver control, add is not applicable (generated before probe).

I'm talking only about udev related information passing needed for systemd.device setup. Other control and monitor goes ofcourse via sysfs, that's out of scope here.

In this context, what would be alternative way of passing such information to udev? I'm not sure that I follow you - do you see as a problem that driver uses "change", or that driver doesnt make sure "bind" carries same information?

This is what driver does:

```
	for (i = 0; i < fan->fan_count; i++) {
		envp[1] = devm_kasprintf(dev, GFP_KERNEL, "INDEX=%d", fan->dt_index[i]);
		if (!envp[1]) {
			result = -ENOMEM;
			goto err_remove_sysfs;
		}

		kobject_uevent_env(&dev->kobj, KOBJ_CHANGE, envp);
	}
```

similar pattern is used throughout kernel source tree (https://elixir.bootlin.com/linux/v5.4.116/A/ident/KOBJ_CHANGE) or "grep -nr kobject_uevent_env.*KOBJ_CHANGE ." if you have source tree at hand.

**Comment by poettering** [comment]
i am sorry, i don't get what this code is supposed to do? you are trying to "stream" information from the kernel to userspace this way? that#s seriously weird, no? what is this "index" thing? why not expose that as sysfs attribute?

**Comment by matijaGP** [comment]
Index is index of fan device on the board, it is same INDEX from this rule:

```
ACTION=="change", KERNEL=="*fsm-fan*", SUBSYSTEM=="platform", ENV{KERNEL}="%k", \
    TAG+="systemd", ENV{SYSTEMD_ALIAS}="/dev/ddal/fan/%E{INDEX}", \
    RUN+="/usr/lib/udev/fan.sh", RUN+="/usr/lib/udev/hwt-fan-fsm.sh"
```

Code is final part of driver's probe function. No streaming of any kind. Code should hit above given rule which is earlier discussed - creation of systemd.device and sysfs-dev symlinks. Everything else (control and monitoring) happens via sysfs, and is done much later, without udev involvement. udev/uevent related are only used to set up device, not for communication.

This is pretty standard, in earlier example block device does similar - sends udev event for userspace to setup whatever needed. Just as block device would share its partition.

**Comment by poettering** [comment]
i really don't follow. partition block device properties are generally also exposing all their attributes from sysfs, and they don#t generate "bind". 

The model of sysfs/udev is that device properties are exposed as sysfs, and optionally also as uevent, but if so then added to every event (with maybe the exception of "remove"), so that userspace can always use them. Properties intrinsic to the event though and not the device as a whole may be associated with one event only.

"tags" are associated with devices, and as long as the same sysfs attrs and event properties are passed down to userspace they will always come to the same conclusion.

It appears to me that you are firing a series of "change" from a single device, that is supposed to be a device property that is supposed to remain sticky, but that's not how the model works, sorry.

Again, is that an upstream driver?

I really don#t get what you are trying do here though.

**Comment by matijaGP** [comment]
Seems like we keep missing each other. Let me try from the other angle. An example from here:

https://www.freedesktop.org/software/systemd/man/systemd.device.html

```
SYSTEMD_READY=
If set to 0, systemd will consider this device unplugged even if it shows up in the udev tree. If this property is unset or set to 1, the device will be considered plugged if it is visible in the udev tree.

This option is useful for devices that initially show up in an uninitialized state in the tree, and for which a "changed" event is generated the moment they are fully set up. Note that SYSTEMD_WANTS= (see above) is not acted on as long as SYSTEMD_READY=0 is set for a device.
```
Is above mentioned change event same change event we talk about here (generated from kernel)? If so, that is as well not part of the sysfs but part of uevent.

Above you said: 'The model of sysfs/udev is that device properties are exposed as sysfs, and optionally also as uevent, but if so then added to every event (with maybe the exception of "remove"), so that userspace can always use them.'

I think this is key in our disagreement - my point is that we cannot assign all properties to each uevent since we do not control all of them. Back to my example, I i.e. driver generates change event and sets property. Next, kernel itself generates uevent (bind), but it will ofcourse not assign properties I assigned earlier, therefore my properties are lost. That's the key element of this report.

**Comment by matijaGP** [comment]
What would simplify this use case, and help to use bind instead of change, is if one could use string substation in ENV, e.g. to say something like

`ENV{SYSTEMD_ALIAS}="/dev/ddal/fan/%s{id}`

unfortunately, env doesnt support it. Is that dictated by something, or simply was never observed necessary?

I do not see possibility to declare systemd_alias from the script as well.

**Comment by arvidjaar** [comment]
@matijaGP 

> Without reloading, everything would work

OK, because I would be rather surprised if it did not - nothing changed in this respect.

> what is your view on this?

I would say systemd view on this is pretty much explained in NEWS and commit messages and you at the end followed it to fix your issue.

> I see it more as a workaround

And what exactly do you suggest? The very reason `bind` events were introduced was that device properties may change. There is no information what properties may be carried over from previous device state. So udev treats each event as completely new.

> Index cannot be read from sysfs 

what exactly prevents your driver from creating `sysfs` attribute for it?

**Comment by matijaGP** [comment]
I dont see Bind to have been introduced because device properties have changed (https://lore.kernel.org/patchwork/patch/811031/). It was addressing problem with Add being fired before probe was completed. So rather to changed properties, it was addressing device state change as some classes of devices will have their state changed in driver probe function.

To have uevent for changed properties, we have Change event. And this is the problem I'm reporting - Bind (which we do not control but is sent by kernel after probe) cleans device properties from previous Change event. That udevadm monitor I posted above is correct timewise.

Where I'm prevented from using index from sysfs (adding is no problem) is inability to use it for setting up alias, string substitution:

ENV{SYSTEMD_ALIAS}="/dev/ddal/fan/%s{index}

is not supported for ENV. I dont see way how ALIAS can be set from the invoked script. Can it? And now that we are on it - wdyt about string substitution support on ENV as well?

**Comment by arvidjaar** [comment]
> string substitution:
> 
> `ENV{SYSTEMD_ALIAS}="/dev/ddal/fan/%s{index}`
> 
> is not supported for `ENV`

Really?

```
$ grep SYSTEMD_ALIAS 90-rdma-umad.rules 
SUBSYSTEM=="infiniband_mad", KERNEL=="*umad*", TAG+="systemd", ENV{SYSTEMD_ALIAS}="/sys/subsystem/rdma/devices/$attr{ibdev}:$attr{port}/umad"
$ 
```

**Comment by yuwata** [comment]
> string substitution:
> 
> ENV{SYSTEMD_ALIAS}="/dev/ddal/fan/%s{index}
> 
> is not supported for ENV.

Maybe missing `"` at the end??

**Comment by poettering** [comment]
> is not supported for ENV. I dont see way how ALIAS can be set from the invoked script. Can it? And now that we are on it - wdyt about string substitution support on ENV as well?

We do support that, as mentioned.

Anyway, I still think there's just a misconception about what uevent properties are: i.e. their lifecycle is specific to the uevent and the time until the next uevent for the device happens. If you want a persistent property, then either make sure to include the property in all uevents, and if you can't do that, then expose it as sysfs attr or suchlike. udev will pick it up from either.

But picking up event properties from one uevent and the propagating them into later uevents is simply conceptually broken. Don't do that.

Anyway, I think we can close this now, no?

**Comment by yuwata** [comment]
Within the discussion, I cannot find any issues in systemd-uded or pid1. This seems to be an issue in the rule or driver.

**Comment by poettering** [comment]
> TAGs are declared sticky, however, they are dropped with next event from current_tags. In device_is_ready, tag is checked in current, not in all. I dont see that to be persistent. Could you clarify please?

tags are a concept for selecting a subset of uevents to subscribe to. If a tag is set on a device and you subscribe uevents with that tag you'll get all current and future uevents for the device, as tags are sticky: once set on a device they stay on the device (in the TAGS field). This is good, because it allows you to not only get the uevents where the tag is set, but also the uevents once they a dropped. i.e. while tags are generally sticky we report in CURRENT_TAG the tags actually applied by the rules that were run on the current uevent. Thus, if you get notified about a uevent that has some tag in TAGS but it's missing in CURRENT_TAGS then this means "hey this device once had the tag assigned in a rule, but in the most recent rule set run it isn't set anymore". And that's what PID1's .device logic monitors and want to know, as it only exposes .device units as active that are around and have the "systemd" tag set in the most recent ruleset run.

or to say this differently: if we wouldn't make tags sticky then this would mean that any uevent that unsets the tags would also mean we'd not propagate the uevent to the watcher, and thus pid 1 would never notice that a device ceased to be relevant, since dropping the tag would then do two things: the device would not be relevant anymore *and* the uevent for it would not be propagated.

> What I also do not quite understand is making use of TAG == systemd to decide on device readiness, that seems conflicting with SYSTEMD_READY ([poettering@242c1c0](https://github.com/poettering/systemd/commit/242c1c075aa284c8a8657c5aca36147f528146ba)). At least, commit message from that commit implies that SYSTEMD_READY should be checked before checking for current tag.

for PID 1 to consider a device up it must show up in udev, have the "systemd" tag currently set and either have SYSTEMD_READY unset or set to 1. PID 1 will consider devices down that are not in udev, or currently don't have a "systemd" tag set, or have SYSTEMD_READY set to 0.

in most cases in your rules you'd set a tag for good, and then finely control with SYSTEMD_READY whether to consider the device active at the current uevent or not. But strictly speaking the latter is kinda redundant, you could instead add/drop the tag alone. The main reason we do both is that tags are primarily a way to subscribe to stuff (i.e they are translated to a bloom filter embedded into the netlink uevent message that allows efficient filtering via BPF), while udev props contain metadata/device state/….  Tracking readiness state exclusively via tags is kind dirty hence. Moreover, tags didn't use to be sticky, and thus back when we added SYSTEMD_READY unsetting a tag would also mean the uevent announcing this would be eaten (see above), and thus it wasn't suitable for marking devices as not ready.

**Comment by poettering** [comment]
Anyway, I think it's clear that this is not a systemd or udev issue really, and there's nothing actionable to fix. Let's hence close.
